plugins/inputs/execd/execd.go

```diff
@@ -0,0 +1,186 @@
+package execd
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/kballard/go-shellquote"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+const sampleConfig = `
+  ## Program to run as daemon
+  command = "telegraf-smartctl"
+
+  ## Define how the process is signaled on each collection interval.
+
+  ## Valid values are:
+  ##   "none"   : Do not signal anything.
+  ##              The process must output metrics by itself.
+  ##   "STDIN"  : Send a newline on STDIN.
+  ##   "SIGHUP" : Send a HUP signal. Not available on Windows.
+  signal = "none"
+
+  ## Data format to consume.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+  data_format = "influx"
+`
+
+type Execd struct {
+	Command string
+	Signal  string
+
+	acc     telegraf.Accumulator
+	cmd     *exec.Cmd
+	parser  parsers.Parser
+	stdin   io.WriteCloser
+	stopped bool
+}
+
+func (e *Execd) SampleConfig() string {
+	return sampleConfig
+}
+
+func (e *Execd) Description() string {
+	return "Run executable as long-running input plugin"
+}
+
+func (e *Execd) SetParser(parser parsers.Parser) {
+	e.parser = parser
+}
+
+func (e *Execd) Start(acc telegraf.Accumulator) error {
+	e.acc = acc
+
+	args, err := shellquote.Split(e.Command)
+	if err != nil || len(args) == 0 {
+		return fmt.Errorf("unable to parse command: %s", err)
+	}
+
+	go e.cmdRun(args)
+
+	return nil
+}
+
+func (e *Execd) Stop() {
+	e.stopped = true
+
+	if e.cmd == nil || e.cmd.Process == nil {
+		return
+	}
+
+	if err := e.cmd.Process.Kill(); err != nil {
+		log.Printf("E! [inputs.execd] FATAL error killing process: %s", err)
+	}
+}
+
+func (e *Execd) cmdRun(args []string) error {
+	var wg sync.WaitGroup
+
+	if len(args) > 1 {
+		e.cmd = exec.Command(args[0], args[1:]...)
+	} else {
+		e.cmd = exec.Command(args[0])
+	}
+
+	stdin, err := e.cmd.StdinPipe()
+	if err != nil {
+		return fmt.Errorf("error opening stdin pipe: %s", err)
+	}
+
+	e.stdin = stdin
+
+	stdout, err := e.cmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf("error opening stdout pipe: %s", err)
+	}
+
+	stderr, err := e.cmd.StderrPipe()
+	if err != nil {
+		return fmt.Errorf("error opening stderr pipe: %s", err)
+	}
+
+	log.Printf("D! [inputs.execd] Start program: %s", e.Command)
+
+	err = e.cmd.Start()
+	if err != nil {
+		return fmt.Errorf("error starting program: %s", err)
+	}
+
+	wg.Add(2)
+
+	go func() {
+		e.cmdReadOut(stdout)
+		wg.Done()
+	}()
+
+	go func() {
+		e.cmdReadErr(stderr)
+		wg.Done()
+	}()
+
+	wg.Wait()
+	e.cmd.Wait()
+
+	if e.stopped {
```

Comment:
I think it would be more straightforward to remove this logic and the 1s delayed restart and just call `e.cmdRun` in `Gather`.

Reply:
Personally I do dislike this as this couples the interval with recovering the process. I do have a few telegraf running with intervals upwards of one hour. They are mainly receiving things.

I can change this nevertheless if there is a solution to the discussion on `func Gather`. What does @danielnelson prefer?

Reply:
Either way works for me, I probably would have done it in Gather but I see your point about how we are using the interval for two functions.

Read my earlier comment about how to stop the process without a race condition first and then what I suggest is to either have the command goroutine sleep at the end of its internal loop or have it read from a channel that is written to by Gather.

If you decide to go with a fixed sleep then I would sleep for 10s, one second is too short IMO.

Reply:
I've made the restart delay configurable in 30498402 with a default of 10 seconds. This should cover all cases. If a daemon needs to be restarted quickly to e.g. not loose events that is possible now.